adam_core/utils/indexable.py

```diff
@@ -1,174 +1,446 @@
-                copy.__dict__[k] = v
-            elif v is None:
-                pass
-            else:
-                err = f"{type(v)} are not supported."
-                raise NotImplementedError(err)
+    @member_index.setter
+    def member_index(self, value):
+        raise ValueError("Cannot set member_index directly. Use set_index() instead.")
 
-        return copy
+    @member_index.deleter
+    def member_index(self):
+        raise ValueError(
+            "Cannot delete member_index. Use set_index() with no arguments instead."
+        )
 
-    def __delitem__(self, i: Union[int, slice, tuple, list, np.ndarray, pd.MultiIndex]):
-        ind = self._handle_index(i)
+    @property
+    def member_length(self):
+        """
+        Length of the class members.
+        """
+        return self._member_length
+
+    @member_length.setter
+    def member_length(self, value):
+        raise ValueError("Cannot set member_length directly. Use set_index() instead.")
 
+    @member_length.deleter
+    def member_length(self):
+        raise ValueError(
+            "Cannot delete member_length. Use set_index() with no arguments instead."
+        )
+
+    def set_index(self, index_values: Optional[Union[str, npt.ArrayLike]] = None):
+        """
+        Set an index on the class for the given values or attribute name.
+
+        Parameters
+        ----------
+        values : str, np.ArrayLike
+            The values to be indexed. If a string is given then the values are taken from the
+            attribute of the same name. If an array is given then the values are taken from the
+            array itself. If None is given then the index is set to the range of length
+            of the class's sliceable members.
+
+        Sets
+        ----
+        self._class_index : `~numpy.ndarray`
+            The externally facing index of the class.
+        self._class_index_to_members : `~numpy.ndarray`
+            The mapping from the externally facing index to the class members.
+        self._class_index_to_members_is_slice : bool
+            Whether the mapping from the externally facing index to the class members are slices.
+        self._class_index_attribute : str, None
+            The attribute on which the index was set, if any. None if no attribute was used.
+        self._member_index : `~numpy.ndarray`
+            The index of the members of the class.
+        self._member_length : int
+            The length of the members of the class.
+
+
+        Raises
+        ------
+        ValueError: If all sliceable members do not have the same length.
+        """
+        # --- Part 1: Determine the validity of this instance's members
+
+        # Scan members and if they are of a type that is sliceable
+        # then store the length. All sliceable members should have the same length
+        # along their first axis.
+        member_lengths = []
+        member_attributes = []
```

Comment:
`member_attributes` is unused, delete it

Reply:
Done.